db/db_iter.cc

```diff
@@ -424,40 +433,66 @@ void DBIter::FindNextUserEntryInternal(bool skipping, bool prefix_check) {
           case kTypeSingleDeletion:
             // Arrange to skip all upcoming entries for this key since
             // they are hidden by this deletion.
-            saved_key_.SetUserKey(
+            // if iterartor specified start_seqnum we
+            // 1) return internal key, including the type
+            // 2) return ikey only if ikey.seqnum >= start_seqnum_
+            // not that if deletion seqnum is < start_seqnum_ we
+            // just skip it like in normal iterator.
+            if (start_seqnum_ > 0 && ikey_.sequence > start_seqnum_)  {
+              saved_key_.SetInternalKey(ikey_);
+              skipping = true;
+              return;
+            } else {
+              saved_key_.SetUserKey(
                 ikey_.user_key,
                 !pin_thru_lifetime_ || !iter_->IsKeyPinned() /* copy */);
-            skipping = true;
-            PERF_COUNTER_ADD(internal_delete_skipped_count, 1);
+                skipping = true;
+                PERF_COUNTER_ADD(internal_delete_skipped_count, 1);
+            }
             break;
           case kTypeValue:
           case kTypeBlobIndex:
-            saved_key_.SetUserKey(
-                ikey_.user_key,
-                !pin_thru_lifetime_ || !iter_->IsKeyPinned() /* copy */);
-            if (range_del_agg_.ShouldDelete(
-                    ikey_, RangeDelAggregator::RangePositioningMode::
-                               kForwardTraversal)) {
-              // Arrange to skip all upcoming entries for this key since
-              // they are hidden by this deletion.
-              skipping = true;
-              num_skipped = 0;
-              PERF_COUNTER_ADD(internal_delete_skipped_count, 1);
-            } else if (ikey_.type == kTypeBlobIndex) {
-              if (!allow_blob_) {
-                ROCKS_LOG_ERROR(logger_, "Encounter unexpected blob index.");
-                status_ = Status::NotSupported(
-                    "Encounter unexpected blob index. Please open DB with "
-                    "rocksdb::blob_db::BlobDB instead.");
-                valid_ = false;
+            if (start_seqnum_ > 0) {
+              // we are taking incremental snapshot here
+              // incremental snapshots aren't supported on DB with range deletes
+              if (ikey_.sequence > start_seqnum_) {
+                saved_key_.SetInternalKey(ikey_);
+                skipping = true;
+                num_skipped = 0;
+                return;
               } else {
-                is_blob_ = true;
-                valid_ = true;
+                // this key and all previous versions shouldn't be included,
```

Comment:
should you call `saved_key_.SetUserKey()` here?

Reply:
I don't think so; the outer

if (start_seqnum_ > 0) {

checks that user requested lower-bounded iterator, and then if we don't get to the branch

if (ikey_.sequence >= start_seqnum_) {

it means that this KV isn't visible at all and should be just skipped. Do i miss anything?

Reply:
I think setting `skipping` alone isn't enough to take the fast-path on L427, which is also necessary to trigger Seek when internal key skips are too many. It checks for both `skipping` and that the `saved_user_key_` hasn't changed yet.

Reply:
Ah, ok, i think i got you - we can update the userkey w/o making it visible to iterator consumer, as long as we don't call `return;` right?

Also it my tests I didn't catch that effect..didn't test how too_many_skipped_rows work with that.

Reply:
@ajkr  btw in the SetInternalKey here I don't bother with pinnable slices etc, unlike we do for user keys, e.g. see

`saved_key_.SetUserKey(
                ikey_.user_key,
                !pin_thru_lifetime_ || !iter_->IsKeyPinned() /* copy */);`

Do you think we need that support for pinnable slices as well here for internal keys?